plugins/pulp_rpm/plugins/db/models.py

```diff
@@ -103,19 +152,55 @@ def __cmp__(self, other):
 
 
 class Distribution(Package):
-    UNIT_KEY_NAMES = ('id', 'family', 'variant', 'version', 'arch')
-    TYPE = ids.TYPE_ID_DISTRO
-
-    def __init__(self, family, variant, version, arch, metadata, id=None):
-        kwargs = locals()
-        # I don't know why this is the "id", but am following the pattern of the
-        # original importer
-        if kwargs['id'] is None:
+    # TODO add docstring to this class
+
+    distribution_id = mongoengine.StringField(required=True)
+    family = mongoengine.StringField(required=True)
+    variant = mongoengine.StringField(required=True)
+    version = mongoengine.StringField(required=True)
+    arch = mongoengine.StringField(required=True)
+
+    files = mongoengine.ListField()
+    timestamp = mongoengine.FloatField()
+    packagedir = mongoengine.StringField()
+
+    # Pretty sure the version_sort_index is never used for Distribution units
+    version_sort_index = mongoengine.StringField()
+
+    # For backward compatibility
+    _ns = mongoengine.StringField(default='units_distribution')
+    _content_type_id = mongoengine.StringField(required=True, default='distribution')
+
+    unit_key_fields = ('distribution_id', 'family', 'variant', 'version', 'arch')
+
+    meta = {'collection': 'units_distribution',
+            'indexes': [
+                'distribution_id', 'family', 'variant', 'version', 'arch',
+                # Unit key Index
+                {
+                    'fields': ['distribution_id', 'family', 'variant', 'version', 'arch'],
+                    'unique': True
+                }],
+            'allow_inheritance': False}
+
+    SERIALIZER = serializers.Distribution
+
+    def __init__(self, *args, **kwargs):
+        """
+        Adds the distribution_id if not already defined, which is derived based on the other 4 unit
+        key fields.
+        """
+        super(Distribution, self).__init__(*args, **kwargs)
+        if not self.distribution_id:
             # the original importer leaves out any elements that are None, so
             # we will blindly trust that here.
-            id_pieces = filter(lambda x: x is not None, ('ks', family, variant, version, arch))
-            kwargs['id'] = '-'.join(id_pieces)
-        super(Distribution, self).__init__(kwargs)
+            id_pieces = filter(lambda x: x is not None,
+                               ('ks',
+                                self.family,
+                                self.variant,
+                                self.version,
+                                self.arch))
+            self.distribution_id = '-'.join(id_pieces)
```

Comment:
Good question. The data model was designed this way long ago before I re-wrote the importer, and I've never heard an explanation.

The reason to not do this in a pre-save signal is that it's very helpful to have the entire unit key available in the instance, for example to do comparisons with other units.